### DISCUSSION TITLE
Running djl in alphine docker image

### DISCUSSION THREAD

**Discussion by rajakumar17** [deepjavalibrary/djl#3503] [Q&A]
I am using langchain4j in a springboot app. The following error happens when the application is dockerized with an alpine linux base image. However, this does not happen with an standard tomcat base image (ubuntu).

Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [dev.langchain4j.model.Tokenizer]: Factory method 'tokenizer' threw exception with message: /root/.djl.ai/tokenizers/0.15.0-0.26.0-linux-x86_64/libtokenizers.so: Error relocating /root/.djl.ai/tokenizers/0.15.0-0.26.0-linux-x86_64/libtokenizers.so: __register_atfork: symbol not found
catalog-svc  | 		at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:177)
catalog-svc  | 		at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:644)
catalog-svc  | 		... 116 more
catalog-svc  | 	Caused by: java.lang.UnsatisfiedLinkError: /root/.djl.ai/tokenizers/0.15.0-0.26.0-linux-x86_64/libtokenizers.so: Error relocating /root/.djl.ai/tokenizers/0.15.0-0.26.0-linux-x86_64/libtokenizers.so: __register_atfork: symbol not found
catalog-svc  | 		at java.base/jdk.internal.loader.NativeLibraries.load(Native Method)
catalog-svc  | 		at java.base/jdk.internal.loader.NativeLibraries$NativeLibraryImpl.open(NativeLibraries.java:388)
catalog-svc  | 		at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:232)
catalog-svc  | 		at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:174)
catalog-svc  | 		at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2394)
catalog-svc  | 		at java.base/java.lang.Runtime.load0(Runtime.java:755)
catalog-svc  | 		at java.base/java.lang.System.load(System.java:1953)
catalog-svc  | 		at ai.djl.huggingface.tokenizers.jni.LibUtils.loadLibrary(LibUtils.java:76)
catalog-svc  | 		at ai.djl.huggingface.tokenizers.jni.LibUtils.<clinit>(LibUtils.java:41)
catalog-svc  | 		at ai.djl.huggingface.tokenizers.HuggingFaceTokenizer.newInstance(HuggingFaceTokenizer.java:173)
catalog-svc  | 		at dev.langchain4j.model.embedding.onnx.HuggingFaceTokenizer.createFrom(HuggingFaceTokenizer.java:92)
catalog-svc  | 		at dev.langchain4j.model.embedding.onnx.HuggingFaceTokenizer.<init>(HuggingFaceTokenizer.java:38)

Native lib location: /root/.djl.ai/tokenizers/0.15.0-0.26.0-linux-x86_64/libtokenizers.so

cat /etc/os-release
NAME="Alpine Linux"

uname -i
unknown

Do you have any clue on how to overcome this issues in alpine image?

**Comment by frankfliu** [comment]
@rajakumar17 Can you try different version of DJL like 0.30.0?

**Reply by rajakumar17** [reply]
I had to add libstdc++ to the docker image. Pos that got the same exception

Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [dev.langchain4j.model.Tokenizer]: Factory method 'tokenizer' threw exception with message: /root/.djl.ai/tokenizers/0.20.0-0.30.0-cpu-linux-x86_64/libtokenizers.so: Error relocating /root/.djl.ai/tokenizers/0.20.0-0.30.0-cpu-linux-x86_64/libtokenizers.so: __register_atfork: symbol not found
catalog-svc  | 		at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:177)
catalog-svc  | 		at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:644)
catalog-svc  | 		... 116 more
catalog-svc  | 	Caused by: java.lang.UnsatisfiedLinkError: /root/.djl.ai/tokenizers/0.20.0-0.30.0-cpu-linux-x86_64/libtokenizers.so: Error relocating /root/.djl.ai/tokenizers/0.20.0-0.30.0-cpu-linux-x86_64/libtokenizers.so: __register_atfork: symbol not found
catalog-svc  | 		at java.base/jdk.internal.loader.NativeLibraries.load(Native Method)
catalog-svc  | 		at java.base/jdk.internal.loader.NativeLibraries$NativeLibraryImpl.open(NativeLibraries.java:388)
catalog-svc  | 		at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:232)
catalog-svc  | 		at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:174)
catalog-svc  | 		at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2394)
catalog-svc  | 		at java.base/java.lang.Runtime.load0(Runtime.java:755)
catalog-svc  | 		at java.base/java.lang.System.load(System.java:1953)
catalog-svc  | 		at ai.djl.huggingface.tokenizers.jni.LibUtils.loadLibrary(LibUtils.java:93)
catalog-svc  | 		at ai.djl.huggingface.tokenizers.jni.LibUtils.<clinit>(LibUtils.java:54)
catalog-svc  | 		at ai.djl.huggingface.tokenizers.HuggingFaceTokenizer.newInstance(HuggingFaceTokenizer.java:185)
catalog-svc  | 		at dev.langchain4j.model.embedding.onnx.HuggingFaceTokenizer.createFrom(HuggingFaceTokenizer.java:92)
catalog-svc  | 		at dev.langchain4j.model.embedding.onnx.HuggingFaceTokenizer.<init>(HuggingFaceTokenizer.java:38)

**Comment by frankfliu** [comment]
@rajakumar17 

I tried on alpine, I got the same error. I did some research. It doesn't seem there is an easy fix. Alpine is `x86_64-alpine-linux-musl` based. the `rustc` compile the target as gnu. I tried to compile tokenizer on alpine, but failed with multiple issues.

Can you consider ubuntu based docker images?
